torch/autograd/_functions/tensor.py

```diff
@@ -565,12 +575,156 @@ def forward(self, input):
+        The way the code calculates it is simply by noting that
+
+        prod_{1 <= i <= j, i != k} x_i
+            = (prod_{1 <= i <= k} x_i) * (prod_{k + 1 <= i <= j} x_i)
+
+        the first term is calculated as prods_until_k, which since
+        doesn't depend in j is easy to vectorize.
+
+        The second term (indexed by j) is the cumulative product of
+        x_{k+1}, x_{k+2}, ..., x_n, and it's named in the code
+        prods_from_k_pkus_1, and it's calculated as a cumprod.
+
+        In order to vectorize this properly, we need to add to
+        omitted_products the dimensions where k > j, and therefore
+        dy_j / dx_k = 0, which is done right after the assert.
+        '''
+
+        input, = self.saved_tensors
+        dim_size = input.size(self.dim)
+        if dim_size == 1:
+            return grad_output.clone()
+
+        #  Simple case with nonzero elements in the input
+        if (input != 0).all():
+            output = torch.cumprod(input, dim=self.dim)
+            return sum_scan_exclusive(output * grad_output, dim=self.dim) / input
+
+        dim_padding = (slice(None, None),) * self.dim
+
+        ones_size = list(input.size())
+        ones_size[self.dim] = 1
+        ones = input.new().resize_(ones_size).fill_(1)
+        zeros = ones * 0
+        grad_input = grad_output.new(input.size()).zero_()
+        for k in range(dim_size):
+            if k == 0:
+                prods_from_k_plus_1 = torch.cumprod(
+                    input[dim_padding + (slice(k + 1, None),)],
+                    dim=self.dim
+                )
+
+                omitted_products = torch.cat(
+                    (ones, prods_from_k_plus_1),
+                    self.dim
+                )
+
+            elif k == dim_size - 1:
+                prods_until_k = torch.prod(
+                    input[dim_padding + (slice(None, k),)],
+                    dim=self.dim
+                )
+
+                omitted_products = prods_until_k
+
+            else:
+                prods_until_k = torch.prod(
+                    input[dim_padding + (slice(None, k),)],
+                    dim=self.dim
+                )
+
+                prods_from_k_plus_1 = torch.cumprod(
+                    input[dim_padding + (slice(k + 1, None),)],
+                    dim=self.dim
+                )
+
+                omitted_products = prods_until_k.expand_as(
+                    prods_from_k_plus_1) * prods_from_k_plus_1
+
+                omitted_products = torch.cat(
+                    (prods_until_k, omitted_products), self.dim)
+
+            # At this point omitted_products is the same size
+            # as input, except on the dimension dim where it's
+            # dim_size - k
+            assert omitted_products.size(self.dim) == dim_size - k
+
+            grad_input.select(self.dim, k).copy_(torch.sum(
+                grad_output[dim_padding + (slice(k, None),)] * omitted_products,
+                dim=self.dim).squeeze())
```

Comment:
Can you remove this `squeeze`? It's going to be unnecessary after the broadcasting PR